Add `add-cookie` and `delete-all-cookies`

plexus opened the pull request:
These seemed useful to have. Especially the second one, as I like to reuse a browser instance across tests, but still want to get back to a clean slate. I decided to add `add-cookie` in one go.

I went with an explicit options map instead of keyword args for `add-cookie`. Don't know what your preference is there. It could also be argued that `add-cookie` should follow the format of `browser-cookies->map`. Just let me know if you want anything changed to `add-cookie`. In that case I'll split up the PR so I can get to that later.

If it would be possible to get another release out that would allow me to demonstrate `page-text` and `delete-all-cookies` in the next Lambda Island episode. At your own convenience and only if you're not too busy of course etc. :)

Thanks a lot! All hail to the SparkleDriver.

jackrusher commented:
Getting a new release out in time for the next Lambda Island should be easy enough!

That said, a couple of small changes to this PR would be very helpful:

I'd rather `add-cookie` be called `set-cookie!`, and that the semantics match that name.

Diving into the jBrowserDriver source code, it looks like `add-cookie` might provide unexpected behavior. Specifically, `(.addCookie (.manage browser) (build-selenium-cookie name value options))` is implemented inside jBD as `cookies.add(cookie)` where `cookies` is `private final Set<Cookie> cookies`. The [contract](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/HashSet.java#HashSet.add%28java.lang.Object%29) for Java's `HashSet` class says of `add` that "If this set already contains the element, the call leaves the set unchanged and returns false." This led me to check `org.openqa.selenium.Cookie`, which [implements](https://seleniumhq.github.io/selenium/docs/api/java/org/openqa/selenium/Cookie.html#equals-java.lang.Object-) `equals` in such a way that "Two cookies are equal if the _name_ and _value_ match". All of which suggests to me that if one were to call `add-cookie` twice with different options, the second call would be silently ignored.

I think this can be fixed easily by adding a call to `(.deleteCookie ...)` before `(.addCookie ...)` within `set-cookie!` to make sure the old one is gone before the new one is "added".

Lastly, I feel like there _should_ be a function to "restore" a set of cookies previously returned by `browser-cookies->map`, but you're probably under a production deadline and it can definitely wait until one of us has time later. :)

plexus commented:
Thanks for digging into that! So to make sure I understood correctly, I should

- rename `add-cookie` to `set-cookie!`
- add a call to `.deleteCookie` at the start of the method

Is the signature of the function fine otherwise?

plexus commented:
I guess I might as well add a `delete-cookie!` while I'm at it?

jackrusher commented:
Oops! I forgot to mention: I'd rather use keyword args only because that's what the browser builder uses and I prefer to be consistent within any given library. Yeah, `delete-cookie!` would be great!

plexus commented:
Alright, I think this is good to go. See what you think.

Deleting the cookie first was slightly more tricky than expected because a naive `(.deleteCookieNamed (.manage browser) name)` deletes cookies with that name on any domain, not just the current or given domain.

    - added delete-cookie!
    - renamed add-cookie to set-cookie!
    - renamed delete-all-cookies to delete-all-cookies!
    - made sure set-cookie! updates the cookie, even when only flags differ
    - made set-cookie! take keyword args instead of an options map
    - made sure all cookie-related bang functions return the browser, for easy threading
    - made it clear in docstrings how functions work when dealing with cookies on multiple domains

jackrusher commented:
Looks good!

jackrusher commented:
0.1.10 is now available at clojars. :)

plexus commented:
Sweet!

plexus commented on test/sparkledriver/core_test.clj:
```diff
@@ -106,7 +106,33 @@
+                 (->> (map str))
+                 sort)
+             ["a-cookie=no domain; path=/bar; domain=0.0.0.0"
+              "a-cookie=some domain; path=/foo; domain=example.org;secure;"
```
Just noticed something here, this should include `HttpOnly`. This seems to be a bug in `org.openqa.selenium.Cookie`. It doesn't add the HttpOnly flag when coercing to string.

https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/Cookie.java#L180-L187

So the flag does get set, and presumably jBrowserDriver respects it, but it doesn't show up here.

@jackrusher
